sdk/pinocchio/src/sysvars/mod.rs

```diff
@@ -44,3 +44,48 @@ macro_rules! impl_sysvar_get {
         }
     };
 }
+
+/// Handler for retrieving a slice of sysvar data from the `sol_get_sysvar`
+/// syscall.
+///
+/// # Safety
+///
+/// `dst` must point to a buffer that can hold at least `length` bytes.
+#[inline]
+pub unsafe fn get_sysvar_unchecked(
+    dst: *mut u8,
+    sysvar_id: &Pubkey,
+    offset: u64,
+    length: u64,
+) -> Result<(), ProgramError> {
+    let sysvar_id = sysvar_id as *const _ as *const u8;
+
+    #[cfg(target_os = "solana")]
+    let result = unsafe { crate::syscalls::sol_get_sysvar(sysvar_id, dst, offset, length) };
+
+    #[cfg(not(target_os = "solana"))]
+    let result = core::hint::black_box(sysvar_id as u64 + dst as u64 + offset + length);
+
+    match result {
+        crate::SUCCESS => Ok(()),
+        e => Err(e.into()),
+    }
+}
+
+/// Handler for retrieving a slice of sysvar data from the `sol_get_sysvar`
+/// syscall.
+#[inline]
+pub fn get_sysvar(
+    dst: &mut [u8],
+    sysvar_id: &Pubkey,
+    offset: u64,
+    length: u64,
+) -> Result<(), ProgramError> {
+    // Check that the provided destination buffer is large enough to hold the
+    // requested data.
+    if dst.len() < length as usize {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    unsafe { get_sysvar_unchecked(dst as *mut _ as *mut u8, sysvar_id, offset, length) }
+}
```

Comment:
`dst` slice already contains the length.
And if someone wants to use a different offset, they can just create a new slice, that points to the desired location.
I think this would be a more canonical use of the slices:

```suggestion
pub fn get_sysvar(
    dst: &mut [u8],
    sysvar_id: &Pubkey,
) -> Result<(), ProgramError> {
    unsafe { get_sysvar_unchecked(dst as *mut _ as *mut u8, sysvar_id, dst.len()) }
}
```

I also do not understand why `offset` is used when `dst` is a pointer.
So I have removed it from the call.  It should just always be `0`.
As one can always create a slice that has an offset like this:

```rust
    let dst_with_offset = &mut dst[offset..];
```

Reply:
It mimics the function from [solana-sdk](https://github.com/anza-xyz/solana-sdk/blob/b0b47b74f1d2b37424598d0cc3f2a7d8bb7fa85b/sysvar/src/lib.rs#L217).
Do you suggest creating a buffer with a size larger than necessary? `offset` can be much larger than the stack size

Reply:
I do not know why the SDK does it this way.
I think this is just the basics of how slices work.
Each slice is a pointer and a length.
You can easily create a new slice from an existing one, with the offset you want.

Why would you need to create a larger buffer?
`dst` is a pointer, not a buffer.
You can create slices from existing slices with an offset, as shown above, and it is a cheap operation.
Which makes the `offset` argument redundant.

I do not see a reason to copy an overly complicated API from the SDK.
But I could be missing something.

Reply:
> The `offset` argument refers to the offset on the sysvar account, so it is required. Agree that we don't need the `length`, which also means we don't need the `get_sysvar_unchecked` variant – we can use `dst.len()` directly.

We probably want an `unchecked` variant to support retrieving data with an uninitialized buffer.

Reply:
Let's do that in a separate PR – we can then decide to either always panic or return dummy values.